crates/goose/src/agents/agent.rs

```diff
@@ -383,15 +383,66 @@ impl Agent {
         sub_recipe_manager.add_sub_recipe_tools(sub_recipes);
     }
 
+    /// Apply schema-aware numeric coercion to tool arguments
+    /// Converts string representations of numbers to actual numbers based on the tool's schema
+    async fn coerce_numeric_arguments(&self, tool_call: &mut CallToolRequestParam) {
+        let Some(ref mut args) = tool_call.arguments else {
+            return;
+        };
+
+        let tools = self.list_tools(None).await;
```

Comment:
we're looking only for one tool, list_tools will actually go and call all mcp servers to ask for their tools - that could be rather expensive! now that I say that, I worry that we are doing this willy-nilly anyway

still though, I think we should move this into the extension manager's dispatch tool call